libtransmission/rpc-server.cc

```diff
@@ -643,6 +643,69 @@ void rpc_server_start_retry_cancel(tr_rpc_server* server)
     server->start_retry_counter = 0;
 }
 
+int tr_evhttp_bind_socket(struct evhttp* httpd, char const* address, ev_uint16_t port)
+{
+#ifdef _WIN32
+    WORD wVersionRequested;
+    WSADATA wsaData;
+    wVersionRequested = MAKEWORD(2, 2);
+    if (WSAStartup(wVersionRequested, &wsaData) != 0)
+    {
+        return evhttp_bind_socket(httpd, address, port);
+    }
+    struct addrinfo* result = NULL;
+    struct addrinfo hints;
+    ZeroMemory(&hints, sizeof(hints));
+    hints.ai_family = AF_UNSPEC;
+    hints.ai_socktype = SOCK_STREAM;
+    hints.ai_protocol = IPPROTO_TCP;
+    hints.ai_flags = AI_PASSIVE;
+
+    if (getaddrinfo(address, std::to_string(port).c_str(), &hints, &result) != 0)
+    {
+        goto CLEANUP_AND_FALLBACK;
+    }
+
+    int fd = socket(result->ai_family, result->ai_socktype, result->ai_protocol);
+    if (fd == INVALID_SOCKET)
+    {
+        freeaddrinfo(result);
+        goto CLEANUP_AND_FALLBACK;
+    }
+    evutil_make_socket_nonblocking(fd);
+    evutil_make_listen_socket_reuseable(fd);
+
+    // Making dual stack
+    if (result->ai_family == AF_INET6)
+    {
+        int off = 0;
+        setsockopt(fd, IPPROTO_IPV6, IPV6_V6ONLY, reinterpret_cast<char*>(&off), sizeof(off));
+    }
+    // Set keep alive
+    int on = 1;
+    setsockopt(fd, SOL_SOCKET, SO_KEEPALIVE, reinterpret_cast<char*>(&on), sizeof(on));
+    if (bind(fd, result->ai_addr, result->ai_addrlen) != 0)
+    {
+        freeaddrinfo(result);
+        goto CLEANUP_SOCKET;
+    }
+    freeaddrinfo(result);
+    if (listen(fd, 128) == -1)
+    {
+        goto CLEANUP_SOCKET;
+    }
+    if (evhttp_accept_socket(httpd, fd) == 0)
+        return 0;
```

Comment:
It also looks to me that we are missing a `WSACleanup()` call if this `if` succeeds and returns. Please implement some sort of cleanup for it.

Reply:
The `result` addrinfo is leaking too in this case.